typescript/sdk/src/ism/HyperlaneIsmFactory.ts

```diff
@@ -0,0 +1,250 @@
+    switch (config.type) {
+      case ModuleType.MULTISIG: {
+        const multisigIsmFactory = this.getContracts(chain).multisigIsmFactory;
+        const address = await this.deployMOfNFactory(
+          chain,
+          multisigIsmFactory,
+          config.validators,
+          config.threshold,
+        );
+        return StaticAggregationIsm__factory.connect(address, signer);
+      }
+      case ModuleType.ROUTING: {
+        const routingIsmFactory = this.getContracts(chain).routingIsmFactory;
+        const isms: ChainMap<types.Address> = {};
+        for (const origin of Object.keys(config.domains)) {
+          const ism = await this.deploy(chain, config.domains[origin]);
+          isms[origin] = ism.address;
+        }
+        const domains = Object.keys(isms).map((chain) =>
+          this.multiProvider.getDomainId(chain),
+        );
+        const modules = Object.values(isms);
+        const tx = await routingIsmFactory.deploy(domains, modules);
+        const receipt = await this.multiProvider.handleTx(chain, tx);
+        const dispatchLogs = receipt.logs
+          .map((log) => {
+            try {
+              return routingIsmFactory.interface.parseLog(log);
+            } catch (e) {
+              return undefined;
+            }
+          })
+          .filter(
+            (log): log is ethers.utils.LogDescription =>
+              !!log && log.name === 'ModuleDeployed',
+          );
+        const moduleAddress = dispatchLogs[0].args['module'];
+        const routingIsm = DomainRoutingIsm__factory.connect(
+          moduleAddress,
+          this.multiProvider.getSigner(chain),
+        );
+        await routingIsm.transferOwnership(config.owner);
+        const address = dispatchLogs[0].args['module'];
+        return DomainRoutingIsm__factory.connect(address, signer);
+      }
+      case ModuleType.AGGREGATION: {
+        const aggregationIsmFactory =
+          this.getContracts(chain).aggregationIsmFactory;
+        const addresses: types.Address[] = [];
+        for (const module of config.modules) {
+          addresses.push((await this.deploy(chain, module)).address);
+        }
+        const address = await this.deployMOfNFactory(
+          chain,
+          aggregationIsmFactory,
+          addresses,
+          config.threshold,
+        );
+        return StaticAggregationIsm__factory.connect(address, signer);
+      }
+      default: {
+        throw new Error('Unknown ModuleType');
+      }
+    }
+  }
+
+  private async deployMOfNFactory(
+    chain: ChainName,
+    factory: StaticMOfNAddressSetFactory,
+    values: types.Address[],
+    threshold: number,
+  ): Promise<types.Address> {
+    const address = await factory.getAddress(values.sort(), threshold);
+    const provider = this.multiProvider.getProvider(chain);
+    const code = await provider.getCode(address);
+    if (code === '0x') {
+      await factory.deploy(values.sort(), threshold);
+    }
```

Comment:
I thought the factory did this protection

Reply:
It does but we would still wind up creating a transaction, no? Better not to submit anything on-chain if we don't have to